bakery-src/scripts/download_exercise_images.py

```diff
@@ -0,0 +1,71 @@
+"""
+Download images from exercises at the end of assemble step.
+Save images into resources folder as checksum filename.
+Replace exercise http links to local resources links.
+"""
+
+import sys
+import requests
+import uuid
+import shutil
+from pathlib import Path
+from lxml import etree
+from tempfile import TemporaryDirectory
+from . import utils
+
+EXERCISE_IMAGE_URL_PREFIX = 'http'
+
+
+def _download_file(url, output_filename):
+    """Download URL content into a file as stream (large file support, low memory)"""
+    # It's possible to optimise it more in future with async parallel downloads?
+    # async download example https://stackoverflow.com/a/73194807/756056
+    with requests.get(url, stream=True) as response:
+        response.raise_for_status()
+        with open(output_filename, 'wb') as out_file:
+            for chunk in response.iter_content(chunk_size=1024*1024):  # 1MB chunks
+                out_file.write(chunk)
+        return output_filename
+
+
+def _download_as_uuid(download_dir, image_url):
+    unique_filename = str(uuid.uuid4())
+    output_filename = Path(download_dir) / Path(unique_filename)
+    _download_file(image_url, str(output_filename))
+    return output_filename
+
+
+def fetch_and_replace_external_exercise_images(resources_dir, input_xml, output_xml):
+    doc = etree.parse(str(input_xml))
+    with TemporaryDirectory() as temp_dir:
```

Comment:
I do not know if it would be worth the trouble, but you could use a [requests.Session](https://requests.readthedocs.io/en/latest/user/advanced/) here and pass it to `_download_as_uuid` and `_download_file`. The session object will try to reuse connections when making requests. It supports the same interface as the requests module, so you can do `session.get(url, stream=True)`, etc. all the same.

Reply:
added it